File: parsers/kotlin/src/main/kotlin/verity/passport/prover/PassportReader.kt
package verity.passport.prover

import org.bouncycastle.asn1.ASN1Sequence
import org.bouncycastle.asn1.pkcs.RSAPublicKey
import verity.passport.prover.Constants.MAX_DG1_SIZE
import verity.passport.prover.Constants.MAX_ECONTENT_SIZE
import verity.passport.prover.Constants.MAX_SIGNED_ATTRIBUTES_SIZE
import verity.passport.prover.Constants.MAX_TBS_SIZE_1300
import verity.passport.prover.Constants.SIG_BYTES
import java.security.MessageDigest

/**
 * All fields extracted from a parsed ePassport required for ZK proof generation.
 * Produced by [PassportReader.extract].
 */
public data class PassportData(
    val dg1Padded: ByteArray,
    val dg1Len: Int,
    val signedAttrs: ByteArray,
    val signedAttributesSize: Int,
    val econtent: ByteArray,
    val econtentLen: Int,
    val dscModulus: ByteArray,
    val dscExponent: Long,
    val dscBarrett: ByteArray,
    val sodSignature: ByteArray,
    val cscaModulus: ByteArray,
    val cscaExponent: Long,
    val cscaBarrett: ByteArray,
    val cscaSignature: ByteArray,
    val country: String,
    val dg1HashOffset: Int,
    val tbsCertificate720: ByteArray,
    val tbsCertificateLen: Int,
    val dscPubkeyOffset: Int,
    val dscExponentOffset: Int,
)

/**
 * Reads and extracts all fields from a parsed ePassport SOD and DG1
 * that are required to generate a ZK proof via the ProveKit pipeline.
 */
public class PassportReader(
    private val dg1: ByteArray,
    private val sod: SOD,
    private val cscaPublicKey: ByteArray? = null
) {

    /**
     * Extracts and assembles all [PassportData] fields needed for ZK proof generation,
     * including padded buffers, moduli, Barrett reductions, and byte offsets.
     */
    public fun extract(): PassportData {
        val dg1Padded = fitBytes(dg1, MAX_DG1_SIZE, "DG1")

        val (signedAttrsPadded, signedAttrsSize) = extractSignedAttrs()
        val (econtentPadded, econtentLen, econtentRaw) = extractEcontent()
        val (dscModulus, dscExponent, dscBarrett, sodSignature) = extractDsc()
        val (cscaModulus, cscaExponent, cscaBarrett, cscaSignature) = extractCsca()

        val dg1Hash = MessageDigest.getInstance("SHA-256").digest(dg1)
        val dg1HashOffset = findOffset(econtentRaw, dg1Hash, "DG1 hash in eContent")

        val country = extractCountry()

        val (tbsCert, tbsCertLen, dscPubkeyOffset) = extractDscCert(dscModulus, MAX_TBS_SIZE_1300)

        // Find the exponent value bytes in TBS (minimal big-endian, matching Rust's find_exponent_offset)
        val tbsBytes = sod.certificate.tbs.bytes
        val dscExponentOffset = findExponentOffset(tbsBytes, dscExponent.toInt(), tbsBytes.size)

        return PassportData(
            dg1Padded = dg1Padded,
            dg1Len = dg1.size,
            signedAttrs = signedAttrsPadded,
            signedAttributesSize = signedAttrsSize,
            econtent = econtentPadded,
            econtentLen = econtentLen,
            dscModulus = dscModulus,
            dscExponent = dscExponent,
            dscBarrett = dscBarrett,
            sodSignature = sodSignature,
            cscaModulus = cscaModulus,
            cscaExponent = cscaExponent,
            cscaBarrett = cscaBarrett,
            cscaSignature = cscaSignature,
            country = country,
            dg1HashOffset = dg1HashOffset,
            tbsCertificate720 = tbsCert,
            tbsCertificateLen = tbsCertLen,
            dscPubkeyOffset = dscPubkeyOffset,
            dscExponentOffset = dscExponentOffset,
        )
    }

    /**
     * Validates DG1 hash consistency against the SOD encapsulated content,
     * and validates the eContent hash against the signed attributes message digest.
     */
    public fun validate() {
        val dg1Hash = MessageDigest.getInstance("SHA-256").digest(dg1)
        val dg1FromEcontent = sod.encapContentInfo.dataGroupHashValues.values[1]
            ?: throw PassportError.MissingDg1Hash()

        if (!dg1Hash.contentEquals(dg1FromEcontent)) {
            throw PassportError.Dg1HashMismatch()
        }

        val econtentHash = MessageDigest.getInstance("SHA-256").digest(sod.encapContentInfo.bytes)
        val msgDigest = sod.signerInfo.signedAttrs.messageDigest.let { raw ->
            if (raw.size > 2 && raw[0] == 0x04.toByte()) raw.copyOfRange(2, raw.size) else raw
        }
        if (!econtentHash.contentEquals(msgDigest)) {
            throw PassportError.EcontentHashMismatch()
        }
    }

    private fun extractSignedAttrs(): SignedAttrsResult {
        val raw = sod.signerInfo.signedAttrs.bytes
        val padded = fitBytes(raw, MAX_SIGNED_ATTRIBUTES_SIZE, "SignedAttributes")
        return SignedAttrsResult(padded, raw.size)
    }

    private fun extractEcontent(): EcontentResult {
        val raw = sod.encapContentInfo.bytes
        val padded = fitBytes(raw, MAX_ECONTENT_SIZE, "eContent")
        return EcontentResult(padded, raw.size, raw)
    }

    private fun extractDsc(): DscFields {
        val pubKeyBytes = sod.certificate.tbs.subjectPublicKeyInfo.subjectPublicKey
        val rsaPubKey = RSAPublicKey.getInstance(ASN1Sequence.fromByteArray(pubKeyBytes))

        val modulus = bigIntToFixedBytes(rsaPubKey.modulus, SIG_BYTES, "DSC modulus")
        val exponent = rsaPubKey.publicExponent.toLong()
        val barrett = BarrettReduction.computeFixed(modulus, SIG_BYTES + 1)
        val signature = fitBytes(sod.signerInfo.signature, SIG_BYTES, "SOD signature")

        return DscFields(modulus, exponent, barrett, signature)
    }

    private fun extractCsca(): CscaFields {
        val cscaSize = SIG_BYTES * 2

        if (cscaPublicKey != null) {
            // The CSCA key may be a full SubjectPublicKeyInfo (from registry JSON)
            // or raw RSA public key bytes. Try SPKI first, fall back to raw.
            val rsaPubKey = try {
                val spki = org.bouncycastle.asn1.x509.SubjectPublicKeyInfo.getInstance(
                    ASN1Sequence.fromByteArray(cscaPublicKey)
                )
                RSAPublicKey.getInstance(spki.parsePublicKey())
            } catch (_: Exception) {
                RSAPublicKey.getInstance(ASN1Sequence.fromByteArray(cscaPublicKey))
            }
            val modulus = bigIntToFixedBytes(rsaPubKey.modulus, cscaSize, "CSCA modulus")
            val exponent = rsaPubKey.publicExponent.toLong()
            val barrett = BarrettReduction.computeFixed(modulus, cscaSize + 1)
            val signature = fitBytes(sod.certificate.signature, cscaSize, "CSCA signature")
            return CscaFields(modulus, exponent, barrett, signature)
        }

        throw PassportError.DataNotFound(
            "CSCA public key not provided. Pass cscaPublicKey to PassportReader constructor."
        )
    }

    private fun extractCountry(): String = extractCountry(dg1)

    private fun extractDscCert(dscModulus: ByteArray, targetSize: Int): DscCertResult {
        val tbsBytes = sod.certificate.tbs.bytes
        val padded = fitBytes(tbsBytes, targetSize, "TBS certificate")
        val pubkeyOffset = findOffset(tbsBytes, dscModulus, "DSC modulus in TBS")
        return DscCertResult(padded, tbsBytes.size, pubkeyOffset)
    }

    /**
     * Companion object exposing utility functions used by the passport proving pipeline.
     */
    public companion object {
        /**
         * Extracts the 3-character ISO country code from the MRZ in DG1.
         */
        public fun extractCountry(dg1: ByteArray): String {
            return if (dg1.size >= 10) {
                String(dg1, 7, 3, Charsets.US_ASCII)
            } else {
                "<<<"
            }
        }

        /**
         * Copies [data] into a zero-padded [ByteArray] of exactly [size] bytes,
         * throwing [PassportError.BufferOverflow] if [data] exceeds [size].
         */
        public fun fitBytes(data: ByteArray, size: Int, label: String): ByteArray {
            if (data.size > size) {
                throw PassportError.BufferOverflow("$label: ${data.size} bytes exceeds buffer $size")
            }
            val result = ByteArray(size)
            data.copyInto(result)
            return result
        }

        /**
         * Converts a [java.math.BigInteger] to a fixed-length big-endian [ByteArray] of [size] bytes,
         * right-aligned with zero padding, stripping any leading sign byte.
         */
        public fun bigIntToFixedBytes(value: java.math.BigInteger, size: Int, label: String): ByteArray {
            var bytes = value.toByteArray()
            if (bytes.size > 1 && bytes[0] == 0.toByte()) {
                bytes = bytes.copyOfRange(1, bytes.size)
            }
            if (bytes.size > size) {
                throw PassportError.BufferOverflow("$label: ${bytes.size} bytes exceeds $size")
            }
            val result = ByteArray(size)
            bytes.copyInto(result, destinationOffset = size - bytes.size)
            return result
        }

        /**
         * Find the offset of the RSA exponent value bytes within TBS certificate.
         * Matches Rust's find_exponent_offset: searches for minimal big-endian bytes.
         */
        public fun findExponentOffset(tbs: ByteArray, exponent: Int, tbsLen: Int): Int {
            val expBe = byteArrayOf(
                ((exponent shr 24) and 0xFF).toByte(),
                ((exponent shr 16) and 0xFF).toByte(),
                ((exponent shr 8) and 0xFF).toByte(),
                (exponent and 0xFF).toByte(),
            )
            // Strip leading zeros (minimal representation)
            val start = expBe.indexOfFirst { it != 0.toByte() }.let { if (it < 0) 3 else it }
            val minimal = expBe.copyOfRange(start, expBe.size)
            return findOffset(tbs.copyOf(tbsLen), minimal, "DSC exponent in TBS")
        }

        /**
         * Searches [haystack] for the first occurrence of [needle], returning its byte offset.
         * Throws [PassportError.DataNotFound] if [needle] is not found.
         */
        public fun findOffset(haystack: ByteArray, needle: ByteArray, label: String): Int {
            outer@ for (i in 0..haystack.size - needle.size) {
                for (j in needle.indices) {
                    if (haystack[i + j] != needle[j]) continue@outer
                }
                return i
            }
            throw PassportError.DataNotFound(label)
        }
    }
}

private data class SignedAttrsResult(
    val padded: ByteArray,
    val size: Int,
)

private data class EcontentResult(
    val padded: ByteArray,
    val length: Int,
    val raw: ByteArray,
)

private data class DscCertResult(
    val tbsCert: ByteArray,
    val tbsCertLen: Int,
    val dscPubkeyOffset: Int,
)

private data class DscFields(
    val modulus: ByteArray,
    val exponent: Long,
    val barrett: ByteArray,
    val signature: ByteArray
)

private data class CscaFields(
    val modulus: ByteArray,
    val exponent: Long,
    val barrett: ByteArray,
    val signature: ByteArray
)

File: parsers/swift/Sources/PassportProverLib/Parsing/PassportReader.swift
import Foundation
import BigInt
import CryptoKit

public struct PassportReader: Sendable {
    private let dg1: Data
    private let sod: SOD
    private let cscaPublicKey: Data?

    public init(dg1: Data, sod: SOD, cscaPublicKey: Data? = nil) {
        self.dg1 = dg1
        self.sod = sod
        self.cscaPublicKey = cscaPublicKey
    }

    // MARK: - Public API

    public func extract() throws -> PassportData {
        let dg1Padded = try dg1.paddedToSize(Constants.maxDg1Size, label: "DG1")

        let (signedAttrs, signedAttrsSize) = try extractSignedAttrs()
        let (econtent, econtentLen, econtentRaw) = try extractEcontent()
        let (dscModulus, dscExponent, dscBarrett, sodSignature) = try extractDsc()
        let (cscaModulus, cscaExponent, cscaBarrett, cscaSignature) = try extractCsca()

        let dg1Hash = Self.sha256(dg1)
        let dg1HashOffset = try Self.findOffset(in: econtentRaw, of: dg1Hash, label: "DG1 hash in eContent")

        let country = extractCountry()

        let (tbsCert, tbsCertLen, dscPubkeyOffset) = try extractDscCert(dscModulus: dscModulus, targetSize: Constants.maxTbsSize1300)

        let tbsBytes = sod.certificate.tbs.bytes
        let dscExponentOffset = try Self.findExponentOffset(tbs: tbsBytes, exponent: Int(dscExponent), tbsLen: tbsBytes.count)

        return PassportData(
            dg1Padded: dg1Padded,
            dg1Len: dg1.count,
            signedAttrs: signedAttrs,
            signedAttributesSize: signedAttrsSize,
            econtent: econtent,
            econtentLen: econtentLen,
            dscModulus: dscModulus,
            dscExponent: dscExponent,
            dscBarrett: dscBarrett,
            sodSignature: sodSignature,
            cscaModulus: cscaModulus,
            cscaExponent: cscaExponent,
            cscaBarrett: cscaBarrett,
            cscaSignature: cscaSignature,
            country: country,
            dg1HashOffset: dg1HashOffset,
            tbsCertificate720: tbsCert,
            tbsCertificateLen: tbsCertLen,
            dscPubkeyOffset: dscPubkeyOffset,
            dscExponentOffset: dscExponentOffset
        )
    }

    public func validate() throws {
        let dg1Hash = Self.sha256(dg1)
        guard let dg1FromEcontent = sod.encapContentInfo.dataGroupHashValues.values[1] else {
            throw PassportError.missingDg1Hash
        }

        if dg1Hash != dg1FromEcontent {
            throw PassportError.dg1HashMismatch
        }

        let econtentHash = Self.sha256(sod.encapContentInfo.bytes)
        var msgDigest = sod.signerInfo.signedAttrs.messageDigest
        // Strip OCTET STRING wrapper if present
        if msgDigest.count > 2, msgDigest[msgDigest.startIndex] == 0x04 {
            msgDigest = Data(msgDigest.dropFirst(2))
        }
        if econtentHash != msgDigest {
            throw PassportError.econtentHashMismatch
        }
    }

    // MARK: - Private Extraction

    private func extractSignedAttrs() throws -> (Data, Int) {
        let raw = sod.signerInfo.signedAttrs.bytes
        let padded = try raw.paddedToSize(Constants.maxSignedAttributesSize, label: "SignedAttributes")
        return (padded, raw.count)
    }

    private func extractEcontent() throws -> (Data, Int, Data) {
        let raw = sod.encapContentInfo.bytes
        let padded = try raw.paddedToSize(Constants.maxEcontentSize, label: "eContent")
        return (padded, raw.count, raw)
    }

    private func extractDsc() throws -> (Data, UInt64, Data, Data) {
        let pubKeyBytes = sod.certificate.tbs.subjectPublicKeyInfo.subjectPublicKey
        let (modulus, exponent) = try parseRSAPublicKey(pubKeyBytes)

        let modulusData = try Self.bigIntToFixedBytes(modulus, size: Constants.sigBytes, label: "DSC modulus")
        let barrett = try BarrettReduction.computeFixed(modulus: modulusData, size: Constants.sigBytes + 1)
        let signature = try sod.signerInfo.signature.paddedToSize(Constants.sigBytes, label: "SOD signature")

        return (modulusData, exponent, barrett, signature)
    }

    private func extractCsca() throws -> (Data, UInt64, Data, Data) {
        let cscaSize = Constants.sigBytes * 2

        guard let cscaKey = cscaPublicKey else {
            throw PassportError.dataNotFound("CSCA public key not provided")
        }

        let (modulus, exponent) = try parseRSAPublicKeyFromSPKI(cscaKey)
        let modulusData = try Self.bigIntToFixedBytes(modulus, size: cscaSize, label: "CSCA modulus")
        let barrett = try BarrettReduction.computeFixed(modulus: modulusData, size: cscaSize + 1)
        let signature = try sod.certificate.signature.paddedToSize(cscaSize, label: "CSCA signature")

        return (modulusData, exponent, barrett, signature)
    }

    private func extractCountry() -> String {
        guard dg1.count >= 10 else { return "<<<" }
        let countryBytes = dg1[dg1.startIndex + 7 ..< dg1.startIndex + 10]
        return String(data: countryBytes, encoding: .ascii) ?? "<<<"
    }

    private func extractDscCert(dscModulus: Data, targetSize: Int) throws -> (Data, Int, Int) {
        let tbsBytes = sod.certificate.tbs.bytes
        let padded = try tbsBytes.paddedToSize(targetSize, label: "TBS certificate")
        let pubkeyOffset = try Self.findOffset(in: tbsBytes, of: dscModulus, label: "DSC modulus in TBS")
        return (padded, tbsBytes.count, pubkeyOffset)
    }

    // MARK: - RSA Key Parsing

    /// Parse RSA public key from raw PKCS#1 RSAPublicKey DER bytes.
    private func parseRSAPublicKey(_ data: Data) throws -> (BigUInt, UInt64) {
        let (node, _) = try ASN1.parse(data)
        let children = try ASN1.parseSequence(node.data)
        guard children.count >= 2 else {
            throw PassportError.invalidDscKey
        }
        let modulus = BigUInt(ASN1.parseIntegerBytes(children[0].data))
        let exponentInt = ASN1.parseIntValue(children[1].data)
        return (modulus, UInt64(exponentInt))
    }

    /// Parse RSA public key from SubjectPublicKeyInfo (SPKI) or raw PKCS#1 format.
    private func parseRSAPublicKeyFromSPKI(_ data: Data) throws -> (BigUInt, UInt64) {
        // Try SPKI first: SEQUENCE { AlgorithmIdentifier, BIT STRING { RSAPublicKey } }
        do {
            let (outerNode, _) = try ASN1.parse(data)
            let children = try ASN1.parseSequence(outerNode.data)
            if children.count >= 2, children[0].tag == ASN1.tagSequence, children[1].tag == ASN1.tagBitString {
                let innerKeyData = DscParser.extractBitStringContent(children[1])
                return try parseRSAPublicKey(innerKeyData)
            }
        } catch {
            // Fall through to try raw format
        }
        return try parseRSAPublicKey(data)
    }

    // MARK: - Byte Search

    public static func findOffset(in haystack: Data, of needle: Data, label: String) throws -> Int {
        guard let range = haystack.range(of: needle) else {
            throw PassportError.dataNotFound(label)
        }
        return haystack.distance(from: haystack.startIndex, to: range.lowerBound)
    }

    public static func findExponentOffset(tbs: Data, exponent: Int, tbsLen: Int) throws -> Int {
        let expBE: [UInt8] = [
            UInt8((exponent >> 24) & 0xFF),
            UInt8((exponent >> 16) & 0xFF),
            UInt8((exponent >> 8) & 0xFF),
            UInt8(exponent & 0xFF),
        ]
        let start = expBE.firstIndex(where: { $0 != 0 }) ?? 3
        let minimal = Data(expBE[start...])
        return try findOffset(in: Data(tbs.prefix(tbsLen)), of: minimal, label: "DSC exponent in TBS")
    }

    // MARK: - Utilities

    public static func bigIntToFixedBytes(_ value: BigUInt, size: Int, label: String) throws -> Data {
        let bytes = Data(value.serialize())
        guard bytes.count <= size else {
            throw PassportError.bufferOverflow("\(label): \(bytes.count) bytes exceeds \(size)")
        }
        var result = Data(count: size)
        let offset = size - bytes.count
        result.replaceSubrange(offset ..< offset + bytes.count, with: bytes)
        return result
    }

    // MARK: - SHA-256

    public static func sha256(_ data: Data) -> Data {
        Data(SHA256.hash(data: data))
    }
}
